transform helper

jryurkanin opened the pull request:
The helper function parses the run properties and results data so that the user doesn't need to include any parsing script in their transform script. The helper function only accepts a user-defined function that accepts an array of results data as its input and returns an array of results data matching the assay design it is used with. Also, the helper function requires that users explicitly define the run properties file path in their transform script, otherwise the labkey transform script parser does not recognize ${runInfo}.

#### Rationale

Users currently must write the parsing script for run properties and results data every time they make a transform script for assays. Because the parsing script will almost always be the same, this helper function can be used to save users time instead.

#### Related Pull Requests

#### Changes
* add helper function for parsing run properties and results-level data in an assay transform script.

jryurkanin committed:
adding helper function to utils.py

The helper function parses the run properties and results data so that the user doesn't need to include any parsing script in their transform script. The helper function only accepts a user-defined function that accepts an array of results data as its input and returns an array of results data matching the assay design it is used with. Also, the helper function requires that users explicitly define the run properties in their transform script, otherwise the labkey transform script parser does not recognize ${runInfo}.

jryurkanin committed:
Create example transform script that uses labkey.utils.transform_helper

jryurkanin committed:
Merge branch 'develop' into fb_helper_func

jryurkanin commented:
i'll do another PR at some point to broaden this so it handles well plates better. right now its dependent on labkey pre-parsing results data in to a tsv.

cnathe reviewed (commented):
These run property type helpers are great! Thank you for pushing this forward. I'll defer to Alan on the actual code review.

labkey-alan commented on samples/assay_transform_function_example.py:
```diff
@@ -0,0 +1,22 @@
+filepath = '${runInfo}'
+
+# define the function that you are using to transform your results data array
+def transform(grid):
```
I am trying to move us away from this samples folder, can you move this code example to a markdown file in the `/docs` folder?

labkey-alan commented on labkey/utils.py:
```diff
@@ -31,3 +31,44 @@ def default(self, o):
+def transform_helper(userTransformFunc, filePathRunProperties):
+    #filePathRunProperties must be explicitly defined as ${runInfo} within the user's transform script
+    #parse run properties to for results data in and out filepaths
+    filePathIn = ''
```
In Python the generally accepted naming convention is to use `snake_case` for variable names, not `camelCase`, and `PascalCase` for classes. Can you convert these variables to be `snake_case`?

labkey-alan reviewed (changes requested):
This is mostly ready, but we should move the example code to the docs folder, and make the code style more Pythonic (and aligned with what we do in this repo).

jryurkanin committed:
Delete assay_transform_function_example.py

moving filter to /docs folder

jryurkanin committed:
documentation for transform_helper function

jryurkanin committed:
Update transform_helper.md

jryurkanin committed:
Update transform_helper.md

jryurkanin committed:
updating transform_helper to be more robust and use snake case

switched from camel case to snake case for variable naming. additionally, changed script to get file_path_in from the row beginning with "runDataUploadedFile" in runProperties file. This is to better handle results data with numerical header names, such as when the results data is a grid representing well plates.

jryurkanin commented:
@labkey-alan okay, should be good now. Let me know if I can make it match our style better... I know I need to be more mindful to keep python scripts pythonic.

labkey-alan commented on labkey/utils.py:
```diff
@@ -31,3 +31,45 @@ def default(self, o):
+    #parse run properties to for results data in and out filepaths
+    file_path_in = ''
+    file_path_out = ''
+    file_run_properties = open(file_path_run_properties) 
+    for l in file_run_properties:
+        row = l.strip().split('\t')
+        if row[0] == 'runDataFile':
+            file_path_out = row[3]
+        if row[0] == 'runDataUploadedFile':
+            file_path_in = row[1]
+    file_run_properties.close()
```
```suggestion
    with open(file_path_run_properties) as file_run_properties:
        for l in file_run_properties:
            row = l.strip().split('\t')
            if row[0] == 'runDataFile':
                file_path_out = row[3]
            if row[0] == 'runDataUploadedFile':
                file_path_in = row[1]
```

labkey-alan replied:
When using open it's best to use it with a `with` statement, that way if any errors happen before close the file handle automatically gets closed.

jryurkanin replied:
got it, makes sense. in retrospect, I've only rarely seen files opened without a with statement.

labkey-alan commented on labkey/utils.py:
```diff
@@ -31,3 +31,45 @@ def default(self, o):
+    file_run_properties.close()
+    
+    #parse results data into array, confirming supported file type is used
+    file_in = open(file_path_in)
```
As above, we should convert this to a with statement

labkey-alan commented on labkey/utils.py:
```diff
@@ -31,3 +31,45 @@ def default(self, o):
+    
+    #write transformed results data array to LabKey assay results data grid
+    #transformed array must be a python list object, not a numpy array or pandas dataframe
+    file_out = open(file_path_out, mode='w')
```
As above we should convert this to a with statement

labkey-alan reviewed (changes requested):
Sorry for the delay in reviewing! Let's clean up these open statements then we can merge and I'll do a release with these changes.

jryurkanin committed:
Update labkey/utils.py

Co-authored-by: Alan Vezina <alanv@labkey.com>

jryurkanin committed:
Update utils.py

using with statements when opening files

jryurkanin committed:
Merge branch 'develop' into fb_helper_func

labkey-alan reviewed (approved):
Looks good!
